File: buysell/src/main/java/piuk/blockchain/androidbuysell/services/BuyConditions.java
package piuk.blockchain.androidbuysell.services;

import info.blockchain.wallet.api.data.Settings;
import info.blockchain.wallet.api.data.WalletOptions;

import io.reactivex.subjects.ReplaySubject;
import piuk.blockchain.androidbuysell.models.ExchangeData;

public class BuyConditions {

    private static BuyConditions instance;

    private ReplaySubject<WalletOptions> walletOptionsSource;
    private ReplaySubject<Settings> walletSettingsSource;
    private ReplaySubject<ExchangeData> exchangeDataSource;

    private BuyConditions(ReplaySubject<WalletOptions> walletOptionsSource,
                          ReplaySubject<Settings> walletSettingsSource,
                          ReplaySubject<ExchangeData> exchangeDataSource) {
        this.walletOptionsSource = walletOptionsSource;
        this.walletSettingsSource = walletSettingsSource;
        this.exchangeDataSource = exchangeDataSource;
    }

    public static BuyConditions getInstance(ReplaySubject<WalletOptions> walletOptionsSubject,
                                            ReplaySubject<Settings> walletSettingsSubject,
                                            ReplaySubject<ExchangeData> coinifyWhitelistedSubject) {
        if (instance == null)
            instance = new BuyConditions(walletOptionsSubject, walletSettingsSubject, coinifyWhitelistedSubject);
        return instance;
    }

    public ReplaySubject<WalletOptions> getWalletOptionsSource() {
        return walletOptionsSource;
    }

    public ReplaySubject<Settings> getWalletSettingsSource() {
        return walletSettingsSource;
    }

    public ReplaySubject<ExchangeData> getExchangeDataSource() {
        return exchangeDataSource;
    }

    public void wipe() {
        instance = null;
    }
}

File: buysell/src/main/java/piuk/blockchain/androidbuysell/services/BuyConditions.kt
package piuk.blockchain.androidbuysell.services

import info.blockchain.wallet.api.data.Settings
import info.blockchain.wallet.api.data.WalletOptions

import io.reactivex.subjects.ReplaySubject
import piuk.blockchain.androidbuysell.models.ExchangeData
import piuk.blockchain.androidcore.utils.annotations.Mockable
import piuk.blockchain.androidcore.utils.helperfunctions.InvalidatableLazy
import javax.inject.Inject
import javax.inject.Singleton

@Mockable
@Singleton
class BuyConditions @Inject constructor() {

    private val optionsInitializer = InvalidatableLazy { ReplaySubject.create<WalletOptions>(1) }
    private val settingsInitializer = InvalidatableLazy { ReplaySubject.create<Settings>(1) }
    private val exchangeInitializer = InvalidatableLazy { ReplaySubject.create<ExchangeData>(1) }

    val walletOptionsSource: ReplaySubject<WalletOptions> by optionsInitializer
    val walletSettingsSource: ReplaySubject<Settings> by settingsInitializer
    val exchangeDataSource: ReplaySubject<ExchangeData> by exchangeInitializer

    fun wipe() {
        optionsInitializer.invalidate()
        settingsInitializer.invalidate()
        exchangeInitializer.invalidate()
    }
}
